File: src/main/java/frc/team449/system/encoder/NEOEncoder.java
package frc.team449.system.encoder;

import com.revrobotics.CANSparkMax;
import com.revrobotics.RelativeEncoder;
import org.jetbrains.annotations.NotNull;

/** A NEO integrated encoder plugged into a Spark */
public final class NEOEncoder extends Encoder {
  public static final int NEO_ENCODER_CPR = 1;

  private final RelativeEncoder enc;

  private NEOEncoder(
      @NotNull String name, @NotNull RelativeEncoder enc, double unitPerRotation, double gearing) {
    super(name, NEO_ENCODER_CPR, 1, 1);
    // Let the underlying encoder do the conversions
    enc.setPositionConversionFactor(unitPerRotation * gearing);
    enc.setVelocityConversionFactor(unitPerRotation * gearing / 60);
    this.enc = enc;
  }

  public static EncoderCreator<CANSparkMax> creator(double unitPerRotation, double gearing) {
    return (name, motor, inverted) ->
        new NEOEncoder(name, motor.getEncoder(), unitPerRotation, gearing);
  }

  @Override
  public double getPositionNative() {
    return enc.getPosition();
  }

  @Override
  public double getVelocityNative() {
    return enc.getVelocity();
  }
}

File: src/main/kotlin/frc/team449/system/encoder/NEOEncoder.kt
package frc.team449.system.encoder

import com.revrobotics.CANSparkMax
import com.revrobotics.RelativeEncoder

/** A NEO integrated encoder plugged into a Spark */
class NEOEncoder(
  name: String,
  private val enc: RelativeEncoder,
  unitPerRotation: Double,
  gearing: Double
) : Encoder(name, NEO_ENCODER_CPR, 1.0, 1.0) {

  init {
    // Let the underlying encoder do the conversions
    enc.setPositionConversionFactor(unitPerRotation * gearing)
    // Divide by 60 because it's originally in RPM
    enc.setVelocityConversionFactor(unitPerRotation * gearing / 60)
  }

  protected override fun getPositionNative() = enc.getPosition()

  protected override fun getVelocityNative(): Double = enc.getVelocity()

  companion object {
    const val NEO_ENCODER_CPR = 1

    fun creator(unitPerRotation: Double, gearing: Double): EncoderCreator<CANSparkMax> = EncoderCreator {
        name, motor, inverted ->
      NEOEncoder(name, motor.getEncoder(), unitPerRotation, gearing)
    }
  }
}
